File: gr00t/Isaac-GR00T/examples/H20/robots/sdk_package/example_py/h10-w/motion/speed_joint.py
import humanoid_sdk_py
import humanoid_sdk_py.h10w as h10w
import time
h10w_motion = h10w.H10wMotion()
h10w_status = h10w.H10wStatus()
h10w_system = h10w.H10wSystem()
time.sleep(1)
print("=== realtime speed joint ===")

ret, move_msg = h10w_status.getMoveMessage()
if ret != 0 or move_msg is None:
    print("获取运动信息失败，任务终止")
    exit(-1)

left_arm_near_zero = all(abs(value - 0.0) < 0.001 for value in move_msg.position[:7])
if not left_arm_near_zero:
    print("为保证安全，请先将左臂回零")
    exit(-1)

ret = h10w_system.powerControl(humanoid_sdk_py.PowerState.POWER_ON)
if ret != 0:
    print("上电失败")
    exit(-1)

ret = h10w_system.clearError()
if ret != 0:
    print("清除错误失败")
    exit(-1)

ret = h10w_system.upperEnableControl(humanoid_sdk_py.EnableState.ENABLE_ON)
if ret != 0:
    print("上半身使能失败")
    exit(-1)


ret = h10w_motion.enableRealtimeCmd(True)
if ret != 0:
    print("打开实时指令失败")
    exit(-1)

vels = h10w.RealtimeJointsParams()
vels.left_arm_valid = False
vels.right_arm_valid = False
vels.torso_valid = False
vels.time = 0.001
for i in range(7):
    vels.left_arm[i] = 0.0

step = 0.0001
time_count = 2000
while time_count > 0:
    if time_count > 1000:
        vels.left_arm[6] += step
    vels.left_arm_valid = True
    ret = h10w_motion.speedJoint(vels)
    if ret != 0:
        print("Speed joint failed.")
        break
    time.sleep(0.001)
    time_count -= 1


ret = h10w_motion.enableRealtimeCmd(False)
if ret != 0:
    print("关闭实时指令失败")
    exit(-1)

left_homing_pos = [
    (h10w.L_ARM_JOINT1, 0.0),
    (h10w.L_ARM_JOINT2, 0.0),
    (h10w.L_ARM_JOINT3, 0.0),
    (h10w.L_ARM_JOINT4, 0.0),
    (h10w.L_ARM_JOINT5, 0.0),
    (h10w.L_ARM_JOINT6, 0.0),
    (h10w.L_ARM_JOINT7, 0.0),
]
ret = h10w_motion.multiJointsMove(left_homing_pos, 0.1, 0.1)
if ret != 0:
    print("Move to zero failed")
    exit(-1)

ret = h10w_motion.waitMove()
if ret != 0:
    print("运动至初始位置超时")
    exit(-1)

File: gr00t/Isaac-GR00T/examples/H20/robots/sdk_package/example/h10-w/motion/speed_joint.cpp
// get_cartesian_rotation_max_acc.cpp

#include <iostream>
#include <iomanip>
#include "h10-w/h10w_motion.h"
#include "common/common_def.h"
#include "common/h10w_def.h"
#include "h10-w/h10w_status.h"
#include "h10-w/h10w_system.h"
#include <algorithm>
#include <thread>

using namespace humanoid;
using namespace humanoid::h10w;
/// @brief
/// 以左臂某端关节为例，将1s内将关节速度从0加速至0.1rad/s，以0.1rad/s的速度运行1s后停止。
int main() {
  std::cout << "=== realtime speed joint ===" << std::endl;
  H10wMotion motion;
  H10wStatus status;
  H10wSystem system;
  MoveMessageParams moveMsg;
  int result = status.getMoveMessage(moveMsg);
  if (result != 0) {
    std::cerr << "获取运动信息失败，任务终止" << std::endl;
    return -1;
  }
  bool leftArmNearZero =
      std::all_of(moveMsg.position.begin(), moveMsg.position.begin() + 7,
                  [](double value) { return value - 0.0 < 0.001; });
  if (!leftArmNearZero) {
    std::cerr << "为保证安全，请先将左臂回零" << std::endl;
    return -1;
  }

  result = system.powerControl(PowerState::POWER_ON);
  if (result != 0) {
    std::cerr << "上电失败" << std::endl;
    return -1;
  }
  result = system.clearError();
  if (result != 0) {
    std::cerr << "清错失败" << std::endl;
    return -1;
  }
  result = system.upperEnableControl(humanoid::EnableState::ENABLE_ON);
  if (result != 0) {
    std::cerr << "上半身使能失败" << std::endl;
    return -1;
  }

  RealtimeJointsParams vels;
  vels.left_arm_valid = false;
  vels.right_arm_valid = false;
  vels.torso_valid = false;
  vels.time = 0.001;

  motion.enableRealtimeCmd(true);
  const int joint_count =
      JointIndexValue::L_ARM_JOINT7 - JointIndexValue::L_ARM_JOINT1 + 1;

  for (int i = 0; i < joint_count; i++) {
    vels.left_arm[i] = 0;
  }

  double step = 0.0001;
  int time = 2000;
  while (time > 0) {
    vels.left_arm[6] += step;
    vels.left_arm_valid = true;
    if (motion.speedJoint(vels) != 0) {
      std::cout << "Speed joint failed." << std::endl;
      break;
    }
    std::this_thread::sleep_for(std::chrono::milliseconds(1));
    time--;
  }
  motion.enableRealtimeCmd(false);

  std::vector<JointTarget> leftZeroPos{
      {static_cast<uint32_t>(JointIndexValue::L_ARM_JOINT7), 0}};
  result = motion.multiJointsMove(leftZeroPos, 0.1, 0.1);
  if (result != 0) {
    std::cout << "Move to zero failed." << std::endl;
  }

  if (motion.waitMove() != 0) {
    std::cerr << "Wait move to zero pos failed." << std::endl;
    return -1;
  }
  return 0;
}
